src/video.h

```diff
@@ -5,186 +5,277 @@
+/** @file video.h Graphics system handling. */
 
-#ifndef VIDEO_H
-#define VIDEO_H
+#ifndef GRAPHICS_H
+#define GRAPHICS_H
 
-#include <memory>
-#include <set>
-#include <SDL.h>
-#include <SDL_ttf.h>
+#include "stdafx.h"
 #include "geometry.h"
 #include "palette.h"
 #include "window_constants.h"
 
+#include <chrono>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+
+#include <GLFW/glfw3.h>
+
+struct FontGlyph;
 class ImageData;
 
-/** Clipped rectangle. */
-class ClippedRectangle {
+using Realtime = std::chrono::high_resolution_clock::time_point;  ///< Represents a time point in real time.
+using Duration = std::chrono::duration<double, std::milli>;       ///< Difference between two time points with millisecond precision.
+
+/**
+ * Get the current real time.
+ * @return The time.
+ */
+inline Realtime Time()
+{
+	return std::chrono::high_resolution_clock::now();
+}
+
+/**
+ * Get the time difference between two time points in milliseconds.
+ * @return The time difference.
+ */
+inline double Delta(const Realtime &start, const Realtime &end = Time())
+{
+	Duration d = end - start;
+	return d.count();
+}
+
+/** Class responsible for rendering text. */
+class TextRenderer {
 public:
-	ClippedRectangle();
-	ClippedRectangle(uint16 x, uint16 y, uint16 w, uint16 h);
-	ClippedRectangle(const ClippedRectangle &cr, uint16 x, uint16 y, uint16 w, uint16 h);
+	static constexpr const uint32 MAX_CODEPOINT = 0xFFFD;  ///< Highest unicode codepoint we can render (arbitrary limit).
 
-	ClippedRectangle(const ClippedRectangle &cr);
-	ClippedRectangle &operator=(const ClippedRectangle &cr);
+	void Initialize();
+	void LoadFont(const std::string &font_path, GLuint font_size);
 
-	void ValidateAddress();
+	GLuint GetTextHeight() const;
+	PointF EstimateBounds(const std::string &text, float scale = 1.0f) const;
 
-	uint16 absx;     ///< Absolute X position in the screen of the top-left.
-	uint16 absy;     ///< Absolute Y position in the screen of the top-left.
-	uint16 width;    ///< Number of columns.
-	uint16 height;   ///< Number of rows.
+	void Draw(const std::string &text, float x, float y, float max_width, uint32 colour, float scale = 1.0f);
 
-	uint32 *address; ///< Base address. @note Call #ValidateAddress prior to use.
-	int32 pitch;     ///< Pitch of a row in bytes. @note Call #ValidateAddress prior to use.
+private:
+	/** Helper struct representing a font glyph. */
+	struct FontGlyph {
+		GLuint texture_id;   ///< The OpenGL texture used to render this glyph.
+		Point16 size;        ///< Size of this glyph in pixels.
+		Point16 bearing;     ///< Alignment offset from the baseline.
+		GLuint advance;      ///< Horizontal spacing.
+		bool valid = false;  ///< If \c false, all data in this struct is invalid.
+	};
+
+	const FontGlyph &GetFontGlyph(const char **text, size_t &length) const;
+
+	FontGlyph characters[MAX_CODEPOINT + 1];  ///< All character glyphs in the current font indexed by their unicode codepoint.
```

Comment:
This could explode in size for foreign languages, not sure how much of a problem that is now (could improve in the future).

Reply:
Currently we only take characters from the lower unicode ranges (ASCII, character variants like äöüñç, some symbols like the °C character) and a few from the very high ranges. The space in-between is already being reserved here so we can index all characters by unicode codepoint, so the glyphs there are not loaded but memory is allocated for them. A lot of foreign scripts actually are in this range so the extra memory consumption will be zero.

The reason why I don't *load* these glyphs though is speed. Loading them all takes about five seconds on startup which is excessive. If/when we get more non-latin-script languages we may want to switch to on-demand loading of character ranges.